File: apps/androidtv/player/engine-mpv/src/main/kotlin/dev/spidola/tv/player/enginempv/MpvLog.kt
// SPDX-FileCopyrightText: 2026 Spidola contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

package dev.spidola.tv.player.enginempv

import android.util.Log
import dev.spidola.tv.core.playercontract.EngineError
import dev.spidola.tv.core.playercontract.PlaybackState
import dev.spidola.tv.core.playercontract.StreamRequest
import dev.spidola.tv.core.playercontract.diagnosticDetail

/**
 * This engine's logcat sink (TECH_SPEC §4.8).
 *
 * The tag matches the core's per-target scheme so one logcat filter shows the whole pipeline
 * from import to playback, and engine state transitions are logged so a support thread has a
 * coherent story.
 *
 * **The hard invariant (§12): no secret value is ever formatted into a log message.** Header
 * values and user-agent tokens are logged by *name only*, never by value, and everything mpv
 * itself says goes through [MpvLogRedaction] first — mpv logs the URL it is opening, and an
 * IPTV URL is routinely made of credentials. There is no log level at which a token becomes
 * loggable, so this file exposes no API that could take one.
 */
internal object MpvLog {
    /** Matches the core's `playback` tracing target (TECH_SPEC §4.8). */
    private const val TAG = "spidola.playback.mpv"

    /** mpv's own log lines, already redacted. Kept at debug: they are verbose and rarely read. */
    fun mpvSaid(
        prefix: String?,
        text: String,
    ) {
        if (!Log.isLoggable(TAG, Log.DEBUG)) return
        Log.d(TAG, "mpv[${prefix.orEmpty()}] ${MpvLogRedaction.redact(text.trimEnd())}")
    }

    /**
     * The request being opened — **names only**.
     *
     * The locator is not logged at all, not even redacted: [MpvLogRedaction] exists to salvage
     * diagnostic value from mpv's output, which we do not control. Here we do control it, and
     * the honest answer is that the shell already knows which channel it asked for, so the URL
     * buys nothing that would justify carrying it past this line.
     */
    fun loading(request: StreamRequest) {
        Log.i(
            TAG,
            "load: buffering=${request.buffering.name}" +
                " headers=${request.headers.joinToString(",") { it.name }.ifEmpty { "none" }}" +
                " userAgent=${if (request.userAgent != null) "overridden" else "default"}",
        )
    }

    /** An engine state transition (TECH_SPEC §4.8: transitions are logged at info/error). */
    fun transition(
        from: PlaybackState,
        to: PlaybackState,
    ) {
        val message = "state: ${from.label} -> ${to.label}"
        if (to is PlaybackState.Failed) {
            // diagnosticDetail is null for every classified variant, so this adds text only
            // for Unknown — which is exactly the case a support thread needs it for.
            Log.e(TAG, "$message (${to.error.logLabel})${to.error.diagnosticDetail?.let { ": $it" }.orEmpty()}")
        } else {
            Log.i(TAG, message)
        }
    }

    fun nativeCallFailed(
        what: String,
        code: Int,
    ) {
        Log.w(TAG, "$what failed: mpv error $code")
    }

    /**
     * The event pump was still running when [MpvClient.release] gave up waiting for it, so the
     * handle was leaked rather than destroyed under a live thread. Nothing here is recoverable;
     * it is logged because it means mpv did not return from `mpv_wait_event` after a wakeup,
     * and a support thread seeing this alongside rising memory has the whole explanation.
     */
    fun pumpOutlivedJoin(timeoutMs: Long) {
        Log.w(TAG, "release: event pump still running after ${timeoutMs}ms — leaking the mpv handle")
    }

    fun nativeLibraryMissing(error: UnsatisfiedLinkError) {
        Log.e(
            TAG,
            "libmpv is not in this build — the mpv engine is unavailable. " +
                "Build it with tools/build-libmpv-android/build.sh.",
            error,
        )
    }

    /** The failure class as a log token: stable, greppable, and not the viewer-facing copy. */
    private val EngineError.logLabel: String
        get() =
            when (this) {
                EngineError.SourceUnreachable -> "SourceUnreachable"
                EngineError.Unauthorized -> "Unauthorized"
                EngineError.UnsupportedFormat -> "UnsupportedFormat"
                EngineError.DecoderFailed -> "DecoderFailed"
                EngineError.Timeout -> "Timeout"
                is EngineError.Unknown -> "Unknown"
            }

    private val PlaybackState.label: String
        get() =
            when (this) {
                PlaybackState.Idle -> "Idle"
                PlaybackState.Loading -> "Loading"
                PlaybackState.Buffering -> "Buffering"
                PlaybackState.Playing -> "Playing"
                PlaybackState.Paused -> "Paused"
                PlaybackState.Ended -> "Ended"
                is PlaybackState.Failed -> "Failed"
            }
}

File: apps/tvos/Packages/PlayerMPV/Sources/PlayerMPV/MPVLog.swift
// SPDX-FileCopyrightText: 2026 Spidola contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

import Foundation
import OSLog
import PlayerContract

extension Logger {
  /// The mpv engine's log category (TECH_SPEC §4.8): one category per span target, so a single
  /// predicate shows the whole playback story across core and engine.
  static let mpv = Logger(subsystem: "dev.spidola.tv", category: "spidola::player::mpv")
}

/// What may be written about a stream, and what may not (TECH_SPEC §12: secret values never
/// appear in interpolated log messages).
///
/// This exists as its own unit because the rule is an invariant, not a habit: every value the
/// engine logs about a request goes through here, so "did we leak a token?" is answered by
/// reading one file rather than auditing every call site.
enum MPVRedaction {
  /// A locator reduced to scheme, host, and port — the parts that diagnose a failure without
  /// carrying a credential.
  ///
  /// Dropping the path is not paranoia. An Xtream locator embeds the account directly in it
  /// (`http://host:8080/live/USERNAME/PASSWORD/123.ts`), and token-bearing query strings are the
  /// norm for the rest, so path and query are exactly where the secrets are. Host and port stay:
  /// they are what make a `sourceUnreachable` report actionable, and they are not credentials.
  static func locatorSummary(_ locator: String) -> String {
    guard let components = URLComponents(string: locator), let scheme = components.scheme else {
      return "<unparsable locator>"
    }
    guard let host = components.host else { return "\(scheme)://<no host>" }
    if let port = components.port { return "\(scheme)://\(host):\(port)" }
    return "\(scheme)://\(host)"
  }

  /// Header names only. A header override exists precisely to carry a token, so its value is
  /// assumed secret with no exception (TECH_SPEC §12) — the names alone answer "which overrides
  /// were applied?", which is the only question the log needs to settle.
  static func headerNames(_ headers: [StreamHeader]) -> String {
    headers.isEmpty ? "<none>" : headers.map(\.name).joined(separator: ", ")
  }

  /// Whether a user-agent override was set — never which one. A UA override is a fingerprint a
  /// source hands out to identify an account, so it is treated as token-shaped.
  static func userAgentPresence(_ userAgent: String?) -> String {
    userAgent == nil ? "default" : "overridden"
  }
}
